homeassistant/components/light/greenwave.py

```diff
@@ -0,0 +1,148 @@
+"""
+Support for TCP Connected (Greenwave Reality) lights.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/light.greenwave/
+"""
+
+import logging
+import voluptuous as vol
+from homeassistant.components.light import (
+    ATTR_BRIGHTNESS, Light, PLATFORM_SCHEMA, SUPPORT_BRIGHTNESS)
+from homeassistant.const import CONF_HOST
+import homeassistant.helpers.config_validation as cv
+SUPPORTED_FEATURES = (SUPPORT_BRIGHTNESS)
+REQUIREMENTS = ['xmltodict==0.11.0']
+_LOGGER = logging.getLogger(__name__)
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST): cv.string,
+})
+
+
+def pull_xml(host):
+    """Pull XML Data from Gateway."""
+    import xmltodict
+    resp = gateway_action(host, 0, 'xml', 0)
+    return xmltodict.parse(resp.content)
+
+
+def number_range(value, left_min, left_max, right_min, right_max):
+    """Map one number range to another."""
+    # Figure out how 'wide' each range is
+    left_span = left_max - left_min
+    right_span = right_max - right_min
+
+    # Convert the left range into a 0-1 range (float)
+    value_scaled = float(value - left_min) / float(left_span)
+
+    # Convert the 0-1 range into a value in the right range.
+    return int(right_min + (value_scaled * right_span))
+
+
+def get_brightness(device):
+    """Get Brightness of Device."""
+    if 'level' in device:
+        level = number_range(int(device['level']), 1, 100, 1, 255)
+        return level
+    else:
+        return 0
+
+
+def is_online(device):
+    """Check Device Status."""
+    return 'offline' not in device
+
+
+def gateway_action(host, did, action, value):
+    """Do Action for TCP Connected Gateway."""
+    import requests
+    if action == 'xml':
+        url = ('http://' + host + '/gwr/gop.php?cmd=GWRBatch&data=<gwrcmds>'

```

Comment:
This is too much protocol specific information. Please extract into a new third party lib.